### ISSUE TITLE
MAX_PATH limitations still impact .NET Core

### ISSUE THREAD

**Issue by nguerrera** [Microsoft/msbuild#4200]
On .NET Core, the managed file system API have some tricks to support long paths irrespective of the application manifest (and machine configuration).

However, on .NET Core

* MSBuild is hosted by dotnet.exe, which does not opt in to long paths in its manifest
* MSBuild does a lot of file system work by P/Invoking straight to Win32

The fixes here are either:

1. Use the x-plat, portable file system code on .NET Core for Windows as on .NET core for other OSes.
2. Get dotnet.exe manifest changed.

(1) would work for all Windows machines running .NET Core msbuild, so that's my preference but I may not understand all the ramifications of doing that.

**Comment by nguerrera** [comment]
Filed https://github.com/dotnet/core-setup/issues/5262 for (2)

**Comment by nguerrera** [comment]
Another approach would be to start using apphosts for the tools in the SDK so that MSBuild.exe can exist on core and have its own manifest. One problem with this is that it breaks running `some/unzipped/path/to/dotnet build` because the apphost won't find a globally installed runtime to use. We considered setting DOTNET_ROOT in the cli in all cases, but that has some drawbacks.

**Comment by rainersigwald** [comment]
I believe that (at the time of the initial port) we found that switching entirely to the managed APIs was a perf hit we weren't willing to take. That is probably better now. There is also the "what subtle behavior differences will ensue?" question; there are some fairly-tricky filesystem operations done via P/Invoke, especially in the `Copy` task.

**Comment by nguerrera** [comment]
cc @danmosemsft @JeremyKuhne

**Comment by JeremyKuhne** [comment]
Perf is definitely not a problem now. I will provide _any_ assistance you need to get off of hand-rolled IO. Together we can avoid much more subtle issues. :) This, for example, has a few:

https://github.com/Microsoft/msbuild/blob/b2db71bebaae4f54f7236ca303e2b0a14aca1a0d/src/Shared/NativeMethodsShared.cs#L1297-L1304

1. You can relatively easily blow out the stack (paths can be 64KB)
2. The length can change from one call to the next, but you're not checking the return value (yay current working directory)
3. In the same vein, it can fail on the second call
4. It won't handle `\\?\` correctly
5. Equality comparison isn't as efficient as it could be
6. Calling twice is expensive when it is _very_ rare that you'll need to- you've already reserved the buffer space

I've also created newer APIs with MSBuild explicitly in mind, notably (in this case):

``` C#
public static string GetFullPath (string path, string basePath);
```

With the idea that MSBuild would use it like `string itemFullPath = GetFullPath(itemInclude, projectFileDirectory);`. That would eliminate the dependency on current working directory- and that API handles device syntax and UNCs correctly, which is non-trivial (e.g. `\\?\UNC\Server\Share` + `..\my\..\..\relative\.\path.txt`.

No matter what happens here I always stand ready to assist with path/io stuff. Don't hesitate to loop me in on issues, PRs, etc. 😁

**Comment by nguerrera** [comment]
> That would eliminate the dependency on current working directory

I don't know if we can ever eliminate this. It's basically a contract to (3rd party) tasks that CWD will be in project directory.

**Comment by JeremyKuhne** [comment]
> I don't know if we can ever eliminate this. It's basically a contract to (3rd party) tasks that CWD will be in project directory. :(

Yeah, you'd still have to set it for compat, but MSBuild itself won't have to depend on it for resolution. Gotta keep crawling along. I'm on a multi-decade quest to mitigate MAX_PATH and Get/SetCurrentDirectory. 😆 

Windows recommends not using [`GetFullPathName()`](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-getfullpathnamew) (and Get/SetCurrentDirectory) but unfortunately doesn't provide an alternative way to resolve relative paths. This new .NET API is part of my answer.

**Comment by SingleAccretion** [comment]
So, I've ran into a similar issue recently, and it was quite a frustrating experience (the resolution required debugging MSBuild source code).

Now, in my case, the problem was caused by the fact the standard generated host (the `bin\Debug\net5.0\Program.exe` one) does not in fact have a manifest setting enabling long path support. This is confirmed with a very simple repro:
```cmd
# The latest SDK, creates a net5.0 project
> dotnet new console -n "LongPathSupport" 
```
```C#
using System;
using System.Runtime.InteropServices;

namespace LongPathSupport
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine($"Long path support is enabled: {RtlAreLongPathsEnabled() != 0}");
        }
		
        [DllImport("ntdll.dll", CallingConvention = CallingConvention.StdCall)]
	public static extern byte RtlAreLongPathsEnabled();
    }
}
```
```cmd
> dotnet build
> bin\Debug\net5.0\LongPathSupport.exe
Long path support is enabled: False
> dotnet bin\Debug\net5.0\LongPathSupport.dll
Long path support is enabled: True
```

In my particular case that meant I was getting strange "could not find the project file `..\..\..\System.Drawing.Common.csproj`" errors when trying to analyze dotnet/runtime with Roslyn Workspaces.

Now, it could be argued that the proper fix would be to have the minihost supplied with a proper manifest (and I will be sure to open an issue about that against the proper repository), but at the same time, this whole thing could've been avoided if MSBuild was using `System.IO` on Windows as it does on other platforms (see `NativeMethodsShared.cs`). There were performance concerns raised with this approach, and they might still be valid, but I also think there definitely is room for improvement as the current experience for people who happen to run into these issues is not good at all.
